bigtable/client/internal/port_platform.h

```diff
@@ -0,0 +1,74 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_PORT_PLATFORM_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_PORT_PLATFORM_H_
+
+/**
+ * @file
+ *
+ * Platform portability details.
+ *
+ * This file discover details about the platform (compiler, OS, and hardware
+ * features) and defines objects and types (mostly macros) to adapt the code.
+ * The file should be fairly minimal because we rely on gRPC to deal with most
+ * of the platform problems.
+ */
+
+// Abort compilation if the compiler does not support C++11.
+#ifndef _MSC_VER
+// Microsoft Visual Studio does not define __cplusplus correctly for C++11.
+#if __cplusplus < 201103L
+#error "Bigtable C++ Client requires C++11"
+#endif  // __cplusplus < 201103L
+#elif _MSC_VER < 1900
+#error "Bigtable C++ Client requires C++11, your version of MSVC is too old"
+#endif  // _MSC_VER
+
+// With Microsoft Visual Studio we need an extra header for the word boolean
+// operators.
+#include <ciso646>
+
+// Discover if exceptions are enabled and define them as needed.
+#ifdef GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#error GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS should not be directly set.
```

Comment:
Would it make sense to keep the `#` flush left, but indent everything after it to show the `ifdef` / `else` / `endif` blocks for readability? I'm having trouble seeing what's inside which block, since there are no curly braces, and no indentation to explain what's going on.

E.g.,

```cpp
#ifdef
#  if defined(...)
    // stuff here
#  else
    // something else here
#  endif
#else
#endif
```

Reply:
Good idea.  Done.

Reply:
Aaaand `clang-format` does not like that.  Bummer.  Thoughts?  I am inclined to do whatever `clang-format` wants.

Reply:
See https://clang.llvm.org/docs/ClangFormatStyleOptions.html; specifically:

> ### IndentPPDirectives (PPDirectiveIndentStyle)

You want the option:

> PPDIS_AfterHash (in configuration: AfterHash) Indents directives after the hash.

Reply:
Unfortunately that is a very recent addition to clang-format.  It is not in the version we use:

* http://releases.llvm.org/4.0.1/tools/clang/ClangFormatStyleOptions.html

Nor in 5.0:

* http://releases.llvm.org/5.0.1/tools/clang/docs/ClangFormatStyleOptions.html

and both of those are newer than whatever is distributed with Ubuntu 16.04 and 14.04.

I have turned off clang-format in this section of the code.  It is sufficiently special that I think that makes the most sense.